File: app/src/main/java/com/krystianwsul/checkme/MyCrashlytics.java
package com.krystianwsul.checkme;

import android.support.annotation.NonNull;
import android.text.TextUtils;
import android.util.Log;

import com.crashlytics.android.Crashlytics;

import junit.framework.Assert;

import io.fabric.sdk.android.Fabric;

public class MyCrashlytics {
    private static Boolean sEnabled = null;

    static void initialize(@NonNull MyApplication myApplication) {
        Assert.assertTrue(sEnabled == null);

        sEnabled = myApplication.getResources().getBoolean(R.bool.crashlytics_enabled);

        if (sEnabled)
            Fabric.with(myApplication, new Crashlytics());
    }

    public static void initialize() {
        Assert.assertTrue(sEnabled == null);

        sEnabled = false;
    }

    public static void log(@NonNull String message) {
        Assert.assertTrue(!TextUtils.isEmpty(message));
        Assert.assertTrue(sEnabled != null);

        Log.e("asdf", "MyCrashLytics.log: " + message);
        if (sEnabled)
            Crashlytics.log(message);
    }

    public static void logException(@NonNull Throwable throwable) {
        Assert.assertTrue(sEnabled != null);

        Log.e("asdf", "MyCrashLytics.logException", throwable);
        if (sEnabled)
            Crashlytics.logException(throwable);
    }

    public static boolean getEnabled() {
        Assert.assertTrue(sEnabled != null);

        return sEnabled;
    }
}

File: app/src/main/java/com/krystianwsul/checkme/MyCrashlytics.kt
package com.krystianwsul.checkme

import android.util.Log
import com.google.firebase.analytics.FirebaseAnalytics
import com.google.firebase.crashlytics.FirebaseCrashlytics
import com.google.firebase.perf.FirebasePerformance
import com.krystianwsul.common.ErrorLogger
import com.krystianwsul.common.utils.UserKey

object MyCrashlytics : ErrorLogger() {

    override val enabled = MyApplication.context
            .resources
            .getBoolean(R.bool.crashlytics_enabled)

    fun init() {
        FirebaseCrashlytics.getInstance().setCrashlyticsCollectionEnabled(enabled)
        FirebaseAnalytics.getInstance(MyApplication.context).setAnalyticsCollectionEnabled(enabled)
        FirebasePerformance.getInstance().isPerformanceCollectionEnabled = enabled

        instance = this
    }

    override fun log(message: String) {
        check(message.isNotEmpty())

        Log.e("asdf", "MyCrashLytics.log: $message")
        if (enabled)
            FirebaseCrashlytics.getInstance().log(message)
    }

    override fun logException(throwable: Throwable) {
        Log.e("asdf", "MyCrashLytics.logException", throwable)
        if (enabled)
            FirebaseCrashlytics.getInstance().recordException(throwable)
    }

    override fun logMethod(obj: Any) {
        val stackTraceElements = Thread.currentThread().stackTrace
        val caller = stackTraceElements[3]

        val method = caller.methodName

        log(obj.javaClass.simpleName + "." + method + " " + obj.hashCode())
    }

    override fun logMethod(obj: Any, message: String) {
        val stackTraceElements = Thread.currentThread().stackTrace
        val caller = stackTraceElements[3]

        val method = caller.methodName

        log(obj.javaClass.simpleName + "." + method + " " + obj.hashCode() + ": $message")
    }

    fun setUserKey(userKey: UserKey) {
        FirebaseCrashlytics.getInstance().setCustomKey("UserKey", userKey.key)
    }
}
